lib/source/Debug.cpp

```diff
@@ -0,0 +1,74 @@
+/**
+ * Copyright (C) 2019 Advanced Micro Devices, Inc. All rights reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell cop-
+ * ies of the Software, and to permit persons to whom the Software is furnished
+ * to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IM-
+ * PLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNE-
+ * CTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include <Tensile/Debug.hpp>
+
+#include <mutex>
+
+#ifndef DEBUG_SM
+#define DEBUG_SM 0
+#endif
+
+namespace Tensile
+{
+    std::once_flag debug_init;
+
+    Debug & Debug::Get()
+    {
+        std::call_once(debug_init, [](){ instance = new Debug(); });
+
+        return *instance;
+    }
+
```

Comment:
It is generally faster to use local `static` variables than to call `call_once`, and it's more concise:

    auto& Debug::Get()
    {
          static auto* instance = new Debug();
          return *instance;
    }

(There's no need for the namespace `debug_init` or `instance` variables anymore.)

In C++11 and later, initialization of local `static` variables is guaranteed to be thread-safe (race-free).

Reply:
Actually, you might not even need dynamic allocation:

    static auto& Debug::Get()
    { 
        static Debug instance;
        return instance;
    }

Reply:
This was written before the LazySingleton class.  If we have both Static and Lazy implemented we can just select which behaviour we want by inheriting from the right parent class.

Reply:
As I said in the other one, a local `static` is faster than using `call_once` and has the same semantics. I'm not talking about changing the semantics; just optimizing the original code to use local `static` to initialize at first use, rather than calling `call_once` or using `new`.

Reply:
Done.